### DISCUSSION TITLE
EMQX webhook limit

### DISCUSSION THREAD

**Discussion by dineshwar21** [emqx/emqx#13703] [Q&A] [answered]
Is there a limit in EMQX webhooks? We receive 4 to 5 publish requests per second, leading to a significant number of messages being processed in the webhook queue. As a result, the EMQX webhook is getting blocked and not sending responses to the webhook.

It's something we need to change config level or any other fix need to make in the webhook handler.

**Comment by id** [comment]
@dineshwar21 how many webhooks do you have, and what is CPU usage on the machine?

**Comment by dineshwar21** [comment]
I'm having only one webhook, with two topic in it. CPU usage seems normal 
![image](https://github.com/user-attachments/assets/d9fe9d78-94d4-4d7b-b2fa-d2bca8f42827)

**Reply by id** [reply]
May be HTTP resource is not responding fast enough?
You could try to tune some of Webhook parameters in the Advanced section, like "Buffer Pool Size", "Request TTL", "Inflight Window".

**Reply by dineshwar21** [reply]
Things I'm doing in http server, log the response and based on a condition inserting data to redis. Normal it take a second or two.

One more thing when so there is 10 to 20 processing message in queue. Node seems to disconnect in webhook. I'm not sure what this for.
![image](https://github.com/user-attachments/assets/832fb122-ff88-4615-94ba-43d95081b91f)

**Reply by dineshwar21** [reply]
@id Is there a buffer size in EMQX that determines how many requests it can handle simultaneously in EMQX

**Reply by id** [reply]
@dineshwar21 take a look at the config parameters I posted above.

**Reply by dineshwar21** [reply]
@id 
![image](https://github.com/user-attachments/assets/bcc59a93-a7e9-4537-b585-c478904c8f27)

This are the information i got in advance config

**Reply by id** [reply]
Correct. Not sure what is exactly the question here?

**Reply by dineshwar21** [reply]
When I'm getting multiple response in processing, my whole webbook get blocked. The Node Status get disconnected. This is something due to buffer queue ? or something else

Also it's there any document for webhook advance setting to find use of Query mode

**Reply by id** [reply, marked as answer]
As I already mentioned, if you see message queued up in the webhook, it means your HTTP server is not responding fast enough. Try to increase "Buffer Pool Size", "Request TTL", "Inflight Window", "Connection Pool Size".

"Async" query mode means webhook can send several requests in parallel and wait for responses asynchronously. In "Sync" query mode each request would block until it gets response from the resource.

**Reply by zmstone** [reply]
Sync mode will generate push-back towards the clients, which should in theory lower the load in EMQX buffers.
However, if you have huge amount of clients doing "sync" calls in parallel, good chance it will still jam the buffer.

A really short "Request TTL" can make the buffer discard requests quickly.

**Reply by dineshwar21** [reply]
Thanks, @id  and @zmstone, for your assistance.
